lib/file-source/src/file_server.rs

```diff
@@ -51,45 +51,63 @@ impl FileServer {
         mut chans: impl Sink<SinkItem = (Bytes, String), SinkError = ()>,
         shutdown: std::sync::mpsc::Receiver<()>,
     ) {
+        let mut read_from_beginning = self.start_at_beginning;
         let mut line_buffer = Vec::new();
         let mut fingerprint_buffer = Vec::new();
 
         let mut fp_map: HashMap<FileFingerprint, FileWatcher> = Default::default();
 
         let mut backoff_cap: usize = 1;
         let mut lines = Vec::new();
-        let mut start_of_run = true;
+
+        let mut checkpointer = Checkpointer::new(&self.data_dir);
+        checkpointer.read_checkpoints(self.ignore_before);
+
         // Alright friends, how does this work?
         //
         // We want to avoid burning up users' CPUs. To do this we sleep after
         // reading lines out of files. But! We want to be responsive as well. We
         // keep track of a 'backoff_cap' to decide how long we'll wait in any
         // given loop. This cap grows each time we fail to read lines in an
-        // exponential fashion to some hard-coded cap.
+        // exponential fashion to some hard-coded cap. To reduce time using glob,
+        // we do not re-scan for major file changes (new files, moves, deletes),
+        // or write new checkpoints, on every iteration.
+        let mut next_glob_time = time::SystemTime::now();
         loop {
-            let mut global_bytes_read: usize = 0;
-            // glob poll
-            let exclude_patterns = self
-                .exclude
-                .iter()
-                .map(|e| Pattern::new(e.to_str().expect("no ability to glob")).unwrap())
-                .collect::<Vec<_>>();
-            for (_file_id, watcher) in &mut fp_map {
-                watcher.set_file_findable(false); // assume not findable until found
-            }
-            for include_pattern in &self.include {
-                for entry in glob(include_pattern.to_str().expect("no ability to glob"))
-                    .expect("Failed to read glob pattern")
-                {
-                    if let Ok(path) = entry {
+            // Glob find files to follow, but not too often.
+            let now_time = time::SystemTime::now();
+            if next_glob_time <= now_time {
+                // Schedule the next glob time.
+                next_glob_time = now_time.checked_add(self.glob_minimum_cooldown).unwrap();
```

Comment:
I did not know that! I think suggested code is functionally equivalent to just replacing SystemTime with Instant.

About leveldb: it looks like it would take more time to understand what to do. Would you want me to add keys to an existing database? A new database seems like overkill for these tiny scraps of data.

Reply:
Oh, good point... for some reason I always imagine `Instant` is limited to `elapsed`. Just swapping it in sounds good! 

And that's fair, it could be a bit much to bring in a whole new db for this. We can go with this as-is for now and keep that in mind as a potential future improvement.